### ISSUE TITLE
Support using other/local LLMs

### ISSUE THREAD

**Issue by DataBassGit** [Significant-Gravitas/AutoGPT#25]
You can modify the code to accept a config file as input, and read the Chosen_Model flag to select the appropriate AI model. Here's an example of how to achieve this:

Create a sample config file named config.ini:

[AI]
Chosen_Model = gpt-4

Offload the call_ai_fuction from the ai_functions.py to a separate library. Modify the call_ai_function function to read the model from the config file:

```
import configparser

def call_ai_function(function, args, description, config_path="config.ini"):
    # Load the configuration file
    config = configparser.ConfigParser()
    config.read(config_path)

    # Get the chosen model from the config file
    model = config.get("AI", "Chosen_Model", fallback="gpt-4")

    # Parse args to comma separated string
    args = ", ".join(args)
    messages = [
        {
            "role": "system",
            "content": f"You are now the following python function: ```# {description}\n{function}```\n\nOnly respond with your `return` value.",
        },
        {"role": "user", "content": args},
    ]

  # Use different AI APIs based on the chosen model
    if model == "gpt-4":
        response = openai.ChatCompletion.create(
            model=model, messages=messages, temperature=0
        )
    elif model == "some_other_api":
        # Add code to call another AI API with the appropriate parameters
        response = some_other_api_call(parameters)
    else:
        raise ValueError(f"Unsupported model: {model}")

    return response.choices[0].message["content"]

```
In this modified version, the call_ai_function takes an additional parameter config_path which defaults to "config.ini". The function reads the config file, retrieves the Chosen_Model value, and uses it as the model for the OpenAI API call. If the Chosen_Model flag is not found in the config file, it defaults to "gpt-4".

 the if/elif structure is used to call different AI APIs based on the chosen model from the configuration file. Replace some_other_api with the name of the API you'd like to use, and replace parameters with the appropriate parameters required by that API. You can extend the if/elif structure to include more AI APIs as needed.

**Comment by Torantulino** [comment]
Excellent work. Lots of people are asking for this submit a pull request!

In order to fully support gpt3.5 (and other models) we need to harden the prompt.

@Koobah had some success by adding this line to the end of prompt.txt:
```
Before submitting the response, simulate parsing the response with Python json.loads. Don't submit unless it can be parsed.
```

This would also help out #21

**Comment by DataBassGit** [comment]
I see you looking over my shoulder. Thoughts?

**Comment by Koobah** [comment]
I also changed the user prompt from NEXT COMMAND to GENERATE NEXT COMMAND JSON. Basically reminding it to use JSON whenever possible
It's still not 100% working, though.

**Comment by DataBassGit** [comment]
https://github.com/DataBassGit/Auto-GPT/blob/master/scripts/ai_function_lib.py

@Koobah This is basically what I'm working with atm. I think we can probably add a verify_json function to a gpt-3.5 segment of that function.

**Comment by DataBassGit** [comment]
Pull request on this is submitted. I'm going to start looking into more models and platforms that can be incorporated.

**Comment by ResourceHog** [comment]
Would be huge if it can run llama.cpp locally.

**Comment by DataBassGit** [comment]
I might be able to swing that. Let's see if this merge gets approved. I'm also looking at implementing GPT4all.

**Comment by MarkSchmidty** [comment]
@DataBassGit I see that PR got closed. What's the status of your fork?

**Comment by DataBassGit** [comment]
They moved the API call to GPT-4 to an external library in main.py, but there are still some scripts that call openai directly, like chat.py, browse.py etc. 

GPT4all doesn't support x64 architecture. I also tried some APIs on hugging face, but it seems that it truncates responses on the free API endpoints.

I just converted BabyAGI to Oobabooga, but it's untested. Should be getting to that tonight. If it works, I will start working on porting AutoGPT to Oobabooga as well. The nice thing about this method is that it allows for local or remote hosting, and can handle many different language models without much issue.

Hugging face should work, though. I need to review Microsoft/Jarvis. They make heavy use of HF APIs.

**Comment by MarkSchmidty** [comment]
GPT4all supports x64 and every architecture llama.cpp supports, which is every architecture (even non-POSIX, and webassemly). Their moto is "Can it run ~Doom~ LLaMA" for a reason.

Ooga supports GPT4all (and all llama.cpp ggml models), since it packages llama.cpp and the llamacpp python bindings library. So porting it to ooba would effectively resolve this.

**Comment by DataBassGit** [comment]
The Python Client for gpt4all only supports x86 Linux and ARM Mac.

**Comment by MarkSchmidty** [comment]
I'm running it on x64 right now.

**Comment by MarkSchmidty** [comment]
To be clear, the x86 architecture for gpt4all should really be called x86/x64. It supports either.

**Comment by alkeryn** [comment]
wouldn't it be simpler to just make an api call to ooba's gui instead of managing the loading of models ?
it may be easier to just have a standardized api so you don't have to care about implementation details.

**Comment by MarkSchmidty** [comment]
Ooba's UI is a lot of overhead just to send and receive requests from a different model.

AFAIK ooba supports two types of models, HuggingFace models and GGML (llama.cpp) models (like GPT4All). The former with HuggingFace libraries and the latter with these python bindings: https://github.com/thomasantony/llamacpp-python for llama.cpp

Adding even basic support for just one of these would surely bring in waves of developers who want local models and who would then contribute to improving Auto-GPT.

**Comment by alkeryn** [comment]
well in the meantime i think i'll fork it to use llama instead, i got gpt4 access but i like the idea of being able to let it run for very long without worrying about cost or api overuse.

**Comment by MarkSchmidty** [comment]
> well in the meantime i think i'll fork it to use llama instead, i got gpt4 access but i like the idea of being able to let it run for very long without worrying about cost or api overuse.

I think a lot of people want this but just don't know it yet. There are lots of interesting use cases which would wrack up a huge OpenAI bill that LLaMA-30B or 65B can probably handle fine for just the cost to power a 150watt $200 Nvidia Tesla P40.

**Comment by MarkSchmidty** [comment]
> well in the meantime i think i'll fork it to use llama instead, i got gpt4 access but i like the idea of being able to let it run for very long without worrying about cost or api overuse.

I think a lot of people want this but just don't know it yet. There are lots of interesting use cases which would wrack up a huge OpenAI bill that LLaMA-30B or 65B can probably handle fine for just the cost to power a 150watt $200 Nvidia Tesla P40.

**Comment by DataBassGit** [comment]
There's an issue with this. Auto-GPT relies on specifically structured prompts in order to function correctly. Llama does not do well at providing prompts structured in the exact format that is required. Vicuna does a much better job. It's not perfect, but could probably get there with some fine tuning.

**Comment by MarkSchmidty** [comment]
With minimal finetuning LLaMA can easily do **better** (yes better\*) than GPT-4. Finetuning goes a long way and LLaMA is a very capable base model. The Vicuna dataset (ShareGPT) is available for finetuning here: [https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/tree/main/HTML_cleaned_raw_dataset](https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/tree/main/HTML_cleaned_raw_dataset)  
  
The ideal finetuning would be based on a dataset of GPT-4's interactions with Auto-GPT though.  
  
\*To be fair, GPT-4 could do better than it already does "out of the box" with a few tweaks like using embeddings, but that is besides the point.

**Comment by DataBassGit** [comment]
@MarkSchmidty We absolutely could port the prompts and responses from autogpt to file and use that for fine tuning vicuna. I don't have a GPU, however, so I'm not able to perform the operation, and I don't have the money for gpt-4.

**Comment by alkeryn** [comment]
@DataBassGit yes this is what i found during trying to implement it, and that's before the pinecone update, after the pinecone update there is an additional use of openai to generate embbedings, which would also need to be made differently.

**Comment by DataBassGit** [comment]
@alkeryn I managed to perform this using sentence_transformers library. This appears to work for Vicuna and pinecone, but you have to change your index dimensions from 1536 to 768 on pinecone. I think the model dictates the index dimensions. I couldn't find a way to adjust the dimensions otherwise.

```from sentence_transformers import SentenceTransformer

model = SentenceTransformer('sentence-transformers/LaBSE')

def get_ada_embedding(text):
    # Get the embedding for the given text
    embedding = model.encode([text])
    return embedding[0]```

**Comment by MarkSchmidty** [comment]
> @alkeryn I managed to perform this using sentence_transformers library. This appears to work for Vicuna and pinecone, but you have to change your index dimensions from 1536 to 768 on pinecone. I think the model dictates the index dimensions. I couldn't find a way to adjust the dimensions otherwise.
> 
> ```
> 
> model = SentenceTransformer('sentence-transformers/LaBSE')
> 
> def get_ada_embedding(text):
>     # Get the embedding for the given text
>     embedding = model.encode([text])
>     return embedding[0]```
> ```

Awesome!

There are offline embedding replacements for pinecone that might be more ideal. For example, https://github.com/wawawario2/long_term_memory is a fork of ooba which produces and stores embeddings locally using zarr and Numpy. See https://github.com/wawawario2/long_term_memory#how-it-works-behind-the-scenes

**Comment by DataBassGit** [comment]
Unfortunately, that would take a lot of chopping to apply to what I am using it for. This is designed for the webui, which I am not using. We're loading ooba in API mode so no --chat or --cai-chat flag. 

python server.py --auto-devices --listen --no-stream

This is how I am initiating the server.

**Comment by MarkSchmidty** [comment]
Right, it would have to be re-implemented specifically for Auto-GPT. I just thought I'd point out that it is a future possibility.

I suppose local embeddings is a separate issue / feature we can look into.

**Comment by DataBassGit** [comment]
I don't expect that @Torantulino will implement local anything. That opens a window for lots of bugs and extra tech debt that he doesn't need. My intention was only to add the capacity for others to replace the api library with one of their own choosing, with the understanding that it's not supported. Thus offloading the api calls to a separate library would give us the ability to build an API interface for whatever we needed, without him needing to support it.

**Comment by MarkSchmidty** [comment]
Neither of us can read @Torantulino's mind.

But if you're right and people want this functionality as much as I suspect they do, either a wave of enthusiastic support for it will sway his mind or the interest will turn into a fork more capable than Auto-GPT (due to the benefits outlined in #348).

**Comment by drusepth** [comment]
Been following this thread while I implement local models in babyagi, but just wanted to pop in and voice my desire to see local models in this project, too. OpenAI is easy to use and implement, but local models have huge benefits in price and and customization which seem paramount to optimize for in projects like these.

**Comment by MarkSchmidty** [comment]
> Been following this thread while I implement local models in babyagi, but just wanted to pop in and voice my desire to see local models in this project, too. OpenAI is easy to use and implement, but local models have huge benefits in price and and customization which seem paramount to optimize for in projects like these.

Local embeddings have these benefits and more as well. If you're on board, check out my meta-feature request at #348 for Fully Local/Offline Auto-GPT :)

**Comment by Torantulino** [comment]
I like the idea of running it offline too, we're looking into it! It would make Auto-GPT that much more accessable.

Thanks for the outstanding interest.

**Comment by 9cento** [comment]
+1

**Comment by DataBassGit** [comment]
I've been working on making an offline port for BabyAGI because it's a much simpler model. One of the issues with offline ports is that each model has a different input format for interpretation. It can also change based on the host api interface. Does it take plain text? JSON?

You also have different token windows per model and different dimensions per memory index. I'm not sure that building a universal agi that can interface with many different models without having to rebuild from scratch is feasible.

**Comment by MarkSchmidty** [comment]
There area already OpenAI API server clones for local models like LLaMA/Alpaca/Vicuna/GPT4All. All you need to do is change the endpoint and context window size. That is not only feasible, it's not even a tall order.

**Comment by DataBassGit** [comment]
@MarkSchmidty I've been looking for one. We've tried llama.cpp, oobabooga, and huggingface. Is there another option you have working with vicuna?

**Comment by alreadydone** [comment]
Maybe [this](https://github.com/iaalm/llama-api-server)? Maybe I have seen one with more stars but I don't remember where.

**Comment by 9cento** [comment]
Are we talking local server or? Otherwise it does make the thing
nonsensical to begin with. Also another layer (local or not) just adds more
latency to the output, theoretically

**Comment by MarkSchmidty** [comment]
> @MarkSchmidty I've been looking for one. We've tried llama.cpp, oobabooga, and huggingface. Is there another option you have working with vicuna?

[https://github.com/alexanderatallah/Alpaca-Turbo](https://github.com/alexanderatallah/Alpaca-Turbo) is a local server implementation of the OpenAI API for Alpaca which can be easily modified for other LLaMA based models. It's a drop-in replacement endpoint for the OpenAI API anywhere the OpenAI API works.  
  
\> Are we talking local server or?  
Yes

**Comment by 9cento** [comment]
Well at least it's something to cope with for the moment, I'll look into it

**Comment by DataBassGit** [comment]
Nice. All three of these look promising. I'll test sometime this week.

**Comment by st4s1k** [comment]
> Would be huge if it can run llama.cpp locally.

* also train it (fine tune it) in parallel with communicating with it (for locally installed LLMs)
* having backup checkpoints

**Comment by 0xgeert** [comment]
Just to add one: Drop in open API replacement as a frontend for Llama CPP. With 400+ stars atm, seems to be one of the more popular: 

https://github.com/abetlen/llama-cpp-python

**Comment by alkeryn** [comment]
@MarkSchmidty wait that's great, if it works we can basically run it on local models with no modifications whatsoever.

**Comment by drusepth** [comment]
Of note, Basaran didn't support Llama-based models until today, but their [newest release](https://github.com/hyperonym/basaran/issues/57#issuecomment-1509036822) finally adds support for them.

**Comment by peakji** [comment]
[Basaran](https://github.com/hyperonym/basaran)'s author here, thanks for recommending our project!

Compared to other LLaMA-specific projects, the advantage of integrating Basaran is that it allows users to freely choose various models from the past and future, not just limited to LLaMA variants. We believe this is particularly important for non-English speaking users, as currently LLaMA is mainly trained on Latin languages.

For compatibility, currently you only need to [modify one environment variable](https://github.com/hyperonym/basaran/blob/master/README.md#openai-client-library) to migrate from the OpenAI text completion API to Basaran, without modifying a single line of code. In the next few weeks, we will also add message template support for models that have undergone instruction tuning, making them fully compatible with the OpenAI Chat API.

**Comment by 9cento** [comment]
> [Basaran](https://github.com/hyperonym/basaran)'s author here, thanks for recommending our project!
> 
> Compared to other LLaMA-specific projects, the advantage of integrating Basaran is that it allows users to freely choose various models from the past and future, not just limited to LLaMA variants. We believe this is particularly important for non-English speaking users, as currently LLaMA is mainly trained on Latin languages.
> 
> For compatibility, currently you only need to [modify one environment variable](https://github.com/hyperonym/basaran/blob/master/README.md#openai-client-library) to migrate from the OpenAI text completion API to Basaran, without modifying a single line of code. In the next few weeks, we will also add message template support for models that have undergone instruction tuning, making them fully compatible with the OpenAI Chat API.

BASEDRAN

**Comment by PurpleStarChild** [comment]
I'm looking at doing something similar to this to make it so oobabooga API is used for accessing a local LLM. I am mostly looking at creating a proof of concept integration, but I'm still learning about the overall code structure of AutoGPT. 

Could someone point me in the direction of the key areas that would need to be touched in order to do this. Or alternatively the areasthat make direct calls to openAI so I know where to start?

**Comment by MarkSchmidty** [comment]
> I'm looking at doing something similar to this to make it so oobabooga API is used for accessing a local LLM. I am mostly looking at creating a proof of concept integration, but I'm still learning about the overall code structure of AutoGPT.
> 
> Could someone point me in the direction of the key areas that would need to be touched in order to do this. Or alternatively the areasthat make direct calls to openAI so I know where to start?

Oogabooga's API is not OpenAI Completions API compatible so it will not work with Auto-GPT. Basarn is an OpenAI Completions API for local models. https://github.com/hyperonym/basaran

**Comment by PurpleStarChild** [comment]
Ahh thank you! I'll give that a try.

**Comment by keldenl** [comment]
i made a drop-in replacement for openai via llama.cpp. almost got it working with autogpt, just need to fully support same-dimension embeddings, if anybody has any pointers pls lmk, i got auto-gpt working for a couple cycles but it's really inconsistent

https://github.com/keldenl/gpt-llama.cpp

**Comment by keldenl** [comment]
> i made a drop-in replacement for openai via llama.cpp (had some issues with the python binding llama-cpp-python). almost got it working with autogpt, just need to fully support same-dimension embeddings, if anybody has any pointers pls lmk, i got auto-gpt working for a couple cycles but it's really inconsistent
> 
> https://github.com/keldenl/gpt-llama.cpp

update: i JUST got autogpt working with llama.cpp! see https://github.com/keldenl/gpt-llama.cpp/issues/2#issuecomment-1514353829

i'm using vicuna for embeddings and generation but it's struggling a bit to generate proper commands to not fall into a infinite loop of attempting to fix itself X(  will look into this tmr but super exciting cuz i got the embeddings working! (turns out it was a bug on my end lol)

had to make some changes to autogpt (add base_url to openai_base_url, and adjust the dimensions of the vector, but otherwise left it alone)

**Comment by Vitorhsantos** [comment]
There is this one too: https://github.com/rhohndorf/Auto-Llama-cpp

We face the same problem here, vicuna understants the comands given via prompt differently from GPT-4, so there is need to re-work the commands.

**Comment by keldenl** [comment]
i wonder if there's another similar agent based package that is designed with gpt3 in mind and shorter prompts that could help with vicuna generation

**Comment by MarkSchmidty** [comment]
StableLM was released today. Has 4096 context window. StableLM 15B will be trained on 1.5T tokens (more than LLaMA)

**Comment by InfernalDread** [comment]
> > i made a drop-in replacement for openai via llama.cpp (had some issues with the python binding llama-cpp-python). almost got it working with autogpt, just need to fully support same-dimension embeddings, if anybody has any pointers pls lmk, i got auto-gpt working for a couple cycles but it's really inconsistent
> > https://github.com/keldenl/gpt-llama.cpp
> 
> update: i JUST got autogpt working with llama.cpp! see [keldenl/gpt-llama.cpp#2 (comment)](https://github.com/keldenl/gpt-llama.cpp/issues/2#issuecomment-1514353829)
> 
> i'm using vicuna for embeddings and generation but it's struggling a bit to generate proper commands to not fall into a infinite loop of attempting to fix itself X( will look into this tmr but super exciting cuz i got the embeddings working! (turns out it was a bug on my end lol)
> 
> had to make some changes to autogpt (add base_url to openai_base_url, and adjust the dimensions of the vector, but otherwise left it alone)

Will you be sharing your fork of AutoGPT so that we can test it out? Thank you by the way for all the work you have done!

**Comment by InfernalDread** [comment]
@keldenl Fantastic! Once again, thank you for all your efforts in making this dream a reality!

**Comment by DGdev91** [comment]
Thanks to @keldenl for his work!

I made a pull request for the changes mentioned by him #2594

**Comment by keldenl** [comment]
awesome stuff @DGdev91 , i'll write up a quick guide tonight on using that fork + gpt-llama.cpp

**Comment by keldenl** [comment]
✨  FULL GUIDE POSTED on how to get `Auto-GPT` running with `llama.cpp` via `gpt-llama.cpp` in https://github.com/keldenl/gpt-llama.cpp/issues/2#issuecomment-1515738173. 

Huge shoutout to @DGdev91  for the PR and hope it gets merged soon!

**Comment by keldenl** [comment]
Quick follow-up, although `gpt-llama.cpp` worked with `auto-gpt` at the time of posting the guide, there were various bugs that plague it from working consistently. I've went ahead and fixed all the ones i could find in yesterday and today, and it's running quite well! (full details here, but it runs continuously forever: https://github.com/keldenl/gpt-llama.cpp/issues/2#issuecomment-1519287093)

now it's down to specifically getting better responses from vicuna :')

**Comment by MarkSchmidty** [comment]
@keldenl Have you heard of SuperCOT? It's a LoRA made for use with LangChain. It may work well merged with Vicuna as an Auto-GPT backend.

**Comment by keldenl** [comment]
@MarkSchmidty  i have not! i'm going to go ahead and try it tonight and report the results here~

**Comment by MarkSchmidty** [comment]
@keldenl The merged versions you can find on HuggingFace are merged with LLaMA. They may perform well. But you can merge with a Vicuna model yourself or keep your eye on this Vicuna/SuperCOT finetune currently in the works: https://huggingface.co/reeducator/vicuna-13b-free/discussions/11#64453a62f993c804b0338fa8

**Comment by keldenl** [comment]
I'm gonna try out the llama supercot version first, and i'll keep an eye on the vicuna version 👀 . thank you! didn't even hear about supercot before this

**Comment by MarkSchmidty** [comment]
I fogot, llama.cpp supports LoRA as of a few days ago. There's a GGML version of the SuperCOT LoRA which can be applied the the Vicuna-13B GGML here: https://huggingface.co/kaiokendev/SuperCOT-LoRA/tree/main/13b/ggml/cutoff-2048

You might try that as well as the LLaMA merged version.

**Comment by 9cento** [comment]
> 

Hi there, quick question: does this work also with more demanding models like vicuna-13B-4bits-128 and gpt4-x-alpaca or is specifically tailored to llama.ccp/.ccp models in general? Thanks

**Comment by 9cento** [comment]
> I forgot, llama.cpp supports applying LoRAs as of a few days ago. There's a GGML version of the SuperCOT LoRA which can be applied to Vicuna-13B at run time here: https://huggingface.co/kaiokendev/SuperCOT-LoRA/tree/main/13b/ggml/cutoff-2048
> 
> You might try that as well as the LLaMA merged version.
Hi there, quick question: does this work also with more demanding models like vicuna-13B-4bits-128 and gpt4-x-alpaca or is specifically tailored to llama.ccp/.ccp models in general? Thanks

**Comment by 9cento** [comment]
> I forgot, llama.cpp supports applying LoRAs as of a few days ago. There's a GGML version of the SuperCOT LoRA which can be applied to Vicuna-13B at run time here: https://huggingface.co/kaiokendev/SuperCOT-LoRA/tree/main/13b/ggml/cutoff-2048
> 
> You might try that as well as the LLaMA merged version.

Hi there, quick question: does this work also with more demanding models like vicuna-13B-4bits-128 and gpt4-x-alpaca or is specifically tailored to llama.ccp/.ccp models in general? Thanks

**Comment by MarkSchmidty** [comment]
There are SuperCOT LoRAs for both GPU and CPU in 30B, 13B, and 7B sizes, as well as merged models for GPU. Some of them are listed in the readme here: https://huggingface.co/kaiokendev/SuperCOT-LoRA

Others are buried in the files here: https://huggingface.co/kaiokendev/SuperCOT-LoRA/tree/main

**Comment by 9cento** [comment]
> There are SuperCOT LoRAs for both GPU and CPU in 30B, 13B, and 7B sizes, as well as merged models for GPU. Some of them are listed in the readme here: https://huggingface.co/kaiokendev/SuperCOT-LoRA You can use the LoRA with any finetune. But they'll work best with finetunes trained with the same prompting format.
> 
> Others are buried in the files here: https://huggingface.co/kaiokendev/SuperCOT-LoRA/tree/main
> 
> For GPU 13B you would want one of these two LoRAs, depending on the cutoff length of the finetune you're using it with: https://huggingface.co/kaiokendev/SuperCOT-LoRA/tree/main/13b/gpu (If you don't know, just try both.)

Sorry but I'm relatively new to this stuff so I'll ask you two more questions. First I just would like to know if your method works virtually for every model (not .ccp only, just to be clear) and then if a LoRA is mandatory or optional. Again, forgive my confusion.

**Comment by MarkSchmidty** [comment]
LoRAs are always optional. They're small files which modify a model. In this case, SuperCOT is a LoRA which modifies a model to make it work with LangChain (and Auto-GPT) better. Any model will work with Auto-GPT. But some work better than others. You can use the SuperCOT LoRAs with any LLaMA based model (including the two you listed). But they're in different formats if they're for cpp or for GPU models. (The different formats are clearly labeled at those links.)

**Comment by Boostrix** [comment]
Also see: #2158 and #25 or #348 / #347

**Comment by manyMachines** [comment]
Documentation for Google's PaLM APIs:

[https://developers.generativeai.google/api](url)

Would be nice to optionally use their embeddings as well in Auto-GPT.

Currently, for preview users of text-bison-001, the input token limit is 8196, output 1024 and is rate limited to 30 requests a minute.

**Comment by lc0rp** [comment]
Linking similar LLM-related comments here and closing them:

GPT4ALL:
- https://github.com/Significant-Gravitas/Auto-GPT/issues/4501
- https://github.com/aorumbayev/autogpt4all (Inspiration)
- https://github.com/aorumbayev/autogpt4all (discussion)

**Comment by alkeryn** [comment]
@mudler neat, someone also mentioned basaran earlier : https://github.com/hyperonym/basaran
so how does it perform in your experience ?

**Comment by Boostrix** [comment]
The ini file is bit funny given how everything else already uses yaml 👍😉

**Comment by mudler** [comment]
> @mudler neat, someone also mentioned basaran earlier : https://github.com/hyperonym/basaran
> so how does it perform in your experience ?

I didn't tried basaran, I've only tried with ggml models and LocalAI. With wizardLM and vicuna-cot seems promising. Not at OpenAI levels, but definitely in the good direction!

**Comment by mudler** [comment]
> The ini file is bit funny given how everything else already uses yaml 👍😉

The env file you mean? You can also put the env variables in the docker compose file and make it a one file only :)

**Comment by isaaclepes** [comment]
I still dream of a day when we can use Petals.dev's API for distributed processing.

**Comment by haochuan-li** [comment]
I'm new to this. I'm wondering if autogpt is able to call custom url(other than openai api) to get response? So that we can use other serving systems like TGI or vllm to serve our own llm.

**Comment by DGdev91** [comment]
> I'm new to this. I'm wondering if autogpt is able to call custom url(other than openai api) to get response? So that we can use other serving systems like TGI or vllm to serve our own llm.

Oh, is this thread still open?

Well, it's possible now. Just set the OPENAI_API_BASE variable and you can use any service wich is compliant with OpenAI's API.

....But local LLMs aren't as good as GPT-4 and i never obtained much, even if is technically possible to use them. So i gave up some time ago.

Maybe some recent long-context LLM like llongma and so on can actually work, but i never tries that.

**Comment by 9cento** [comment]
> > I'm new to this. I'm wondering if autogpt is able to call custom url(other than openai api) to get response? So that we can use other serving systems like TGI or vllm to serve our own llm.
> 
> Oh, is this thread still open?
> 
> Well, it's possible now. Just set the OPENAI_API_BASE variable and you can use any service wich is compliant with OpenAI's API.
> 
> ....But local LLMs aren't as good as GPT-4 and i never obtained much, even if is technically possible to use them. So i gave up some time ago.
> 
> Maybe some recent long-context LLM like llongma and so on can actually work, but i never tried that.

Did you gave a try to Llama 2? Codellama?

**Comment by DataBassGit** [comment]
The issue with open source models is that they are trained differently. Getting a response in a specific format requires either fine tuning of the model or modification of the prompts. AutoGPT wasn't designed to make it easy to edit the prompts, and fine tuning is expensive. Eventually, I just built my own agent framework.

**Comment by chymian** [comment]
One Proxy  to rule them all!

https://github.com/BerriAI/litellm/

is a API-Proxy with a vast choice on backends, like replicat, openai, petals, ...
and it works like charm.
pls implement!

**Comment by Wladastic** [comment]
> One Proxy to rule them all!
> 
> https://github.com/BerriAI/litellm/
> 
> is a API-Proxy with a vast choice on backends, like replicat, openai, petals, ... and it works like charm. pls implement!

Thank you for your suggestion,
I will take a look at this.

**Comment by alkeryn** [comment]
@Wladastic just so you know textgen webui has a openai like api now, you can look in the wiki on how to force openai to use its api.

**Comment by Wladastic** [comment]
I know about that one,
I was talking about the project mentioned above

**Comment by Pwuts** [comment]
Coming soon... :)

**Comment by GoZippy** [comment]
I got your activity

**Comment by alkeryn** [comment]
i mean at that point it is pretty easy to tell it to use other OpenAPI compatible backends with the env variables.
and there must be proxies to use other providers too ie convert a non OpenAPI api to OpenAPI.

**Comment by ntindle** [comment]
We've made lots of progress on this front recently with @Pwuts 's work on additional providers. We can now begin the work on more open providers. We will likely be starting with llamafile then moving out from there
